### PR TITLE
Changed unstack (pivot wider) dialog to use pivot_wider function

### LINKED ISSUE #6871: Change unstack to use pivot_wider
This is the partner to the stack (pivot longer) that has been changed recently.

It should be easy.  The Factor to unstack by is the names_from argument
The variable to unstack is values_from
The carry columns is id_cols.
However, the default is to carry all remaining variables, so if there are no carry variables, then I guess id_cols should be empty?

There are 2 small, but useful additions
A checkbox, default unchecked with label Add Prefix to Variable Names.  If checked it opens a field into which a text can be typed (adds the argument manes_prefix
A checkbox, default unchecked with label Fill Missing Values.  If checked it opens a field into which a number can be typed, default is 0. (adds the argument values_fill)

### PR DESCRIPTION
**Pull request by Gioche6**
Fixes #6871

@africanmathsinitiative/developers , This is ready for review
@shadrackkibet , I have changed the function from reshape to `pivot_wide`r as suggested. Kindly review if the codes works well. 
@rdstern , I have added the `values_fill` and `names_prefix `checkboxes as suggested in issue #6871. Kindly test if it's now alright.

### CODE REVIEW

**Comment by Ivanluv** [comment]
@Gioche6  is this ready for review

**Review by rdstern — DISMISSED** [review]
@Gioche6 great.  From my point of view this seems to work.  Let's see what @shadrackkibet says

**Comment by Gioche6** [comment]
@rdstern, Currently  I am working on implementing the code nicely to prevent any future bugs with the help of @Ivanluv . I will push the changes once done for testing. For  now, this is still in progress. Thank you

**Comment by rdstern** [comment]
@Gioche6 thank you for the update.  Let me know when I can check again.

**Comment by Gioche6** [comment]
@rdstern . This is ready for testing Now. I have been working on how to best implement the carry columns control and the multiple receiver for columns to unstack. I believe it works better now. I look forward to your comments on this. 

@Ivanluv , Kindly review the same. Thank you

File: instat/dlgUnstack.vb [hunk 1]
```diff
@@ -41,119 +43,211 @@ Public Class dlgUnstack
...
+        clsDcastFunction.SetRCommand("pivot_wider")
+
+        clsHierachyFunction.SetPackageName("reshape2")
+        clsHierachyFunction.SetRCommand("dcast")
+        clsHierachyFunction.AddParameter("data", clsRFunctionParameter:=ucrSelectorForUnstack.ucrAvailableDataFrames.clsCurrDataFrame, iPosition:=0)
+        clsHierachyFunction.AddParameter("formula", clsROperatorParameter:=clsFormula, iPosition:=1)
 
         clsSelectFunction.SetPackageName("dplyr")
         clsSelectFunction.SetRCommand("select")
         clsSelectFunction.AddParameter("comma", clsROperatorParameter:=clsCommaOperator, bIncludeArgumentName:=False)
```
Comment (Ivanluv) [inline instat/dlgUnstack.vb:161]: parameter position

File: instat/dlgUnstack.vb [hunk 2]
```diff
@@ -41,119 +43,211 @@ Public Class dlgUnstack
...
-        ucrReceiverCarryColumns.SetRCode(clsCommaOperator, bReset)
-        ucrChkDropMissingCombinations.SetRCode(clsDcastFunction, bReset)
-        ucrReceiverFactorToUnstackby.SetRCode(clsFormula, bReset)
     End Sub
 
-    Private Sub TestOKEnabled()
-        If Not ucrReceiverFactorToUnstackby.IsEmpty() AndAlso ucrNewDFName.IsComplete Then
-            ucrBase.OKEnabled(True)
+    Private Sub ucrMultipleColumnsReceiver_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrMultipleColumnsReceiver.ControlValueChanged
+
```
Comment (Ivanluv) [inline instat/dlgUnstack.vb:189]: remove the space

File: instat/dlgUnstack.vb [hunk 3]
```diff
@@ -41,119 +43,211 @@ Public Class dlgUnstack
...
+        ValuesfillParameter()
+    End Sub
+
+    Private Sub ucrChkCarryColumns_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkCarryColumns.ControlValueChanged
+        AddRemoveDataOrPipeOperator()
+        CarryColumnsParameter()
+
```
Comment (Ivanluv) [inline instat/dlgUnstack.vb:196]: Remove the space

File: instat/dlgUnstack.vb (hunk 2 above)
Comment (Gioche6) [reply instat/dlgUnstack.vb]: Done

File: instat/dlgUnstack.vb (hunk 3 above)
Comment (Gioche6) [reply instat/dlgUnstack.vb]: Done

**Comment by Gioche6** [comment]
@Ivanluv , I've done the changes as you requested.

**Review by Ivanluv — DISMISSED** [review]

**Comment by Gioche6** [comment]
@rdstern , This is now ready for testing. Thank you

**Review by rdstern — APPROVED** [review]
@Gioche6 - great - seems to work fine.

File: instat/dlgUnstack.vb [hunk 4]
```diff
@@ -41,119 +43,209 @@ Public Class dlgUnstack
...
         clsSelectFunction = New RFunction
+        clsHierachyFunction = New RFunction
         clsUnstackedOperator = New ROperator
         clsCommaOperator = New ROperator
-        clsFormula = New ROperator
         clsCarryColumnsOperator = New ROperator
+        clsFormula = New ROperator
+        clsDummyFunction = New RFunction
+
+        ucrChkCarryColumns.Checked = False
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:129]: Remove the manual checking. You could use a dummy function to set the default.

File: instat/dlgUnstack.vb [hunk 5]
```diff
@@ -19,20 +19,22 @@ Imports instat.Translations
...
     Private bReset As Boolean = True
-    Private clsFormula As New ROperator
     Private clsCarryColumnsOperator As New ROperator
     Private clsSelectFunction As New RFunction
-    Private clsUnstackedOperator, clsCommaOperator As New ROperator
+    'clsDummyFunction controls the state of the UcrpnlUnstackCol 
+    Private clsDummyFunction As New RFunction
+    Private clsHierachyFunction As New RFunction
+    Private clsSelectDataFunction As New RFunction
+    Private clsUnstackedOperator, clsCommaOperator, clsFormula, clspipeOperator As New ROperator
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:28]: `clsFormulaOperator`?

File: instat/dlgUnstack.vb [hunk 6]
```diff
@@ -189,39 +285,89 @@ Public Class dlgUnstack
...
     End Sub
 
-    Private Sub ucrCoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrNewDFName.ControlContentsChanged, ucrReceiverFactorToUnstackby.ControlContentsChanged, ucrReceiverCarryColumns.ControlContentsChanged
+    Private Sub ucrCoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrNewDFName.ControlContentsChanged, ucrReceiverFactorToUnstackby.ControlContentsChanged, ucrReceiverCarryColumns.ControlContentsChanged, ucrNudValuesFill.ControlContentsChanged, ucrInputTextPrefix.ControlContentsChanged, ucrChkAddPrefix.ControlContentsChanged, ucrChkValuesFill.ControlContentsChanged, ucrMultipleColumnsReceiver.ControlContentsChanged, ucrReceiverColumnToUnstack.ControlContentsChanged
         TestOKEnabled()
     End Sub
+
+    Private Sub ucrChkValuesFill_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkValuesFill.ControlValueChanged
+        If ucrChkValuesFill.Checked Then
+            clsDcastFunction.AddParameter(New RParameter("values_fill", "0", iNewPosition:=7))
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:366]: Technically we should be doing this under `InitialiseDialog()`. I am keen to understand why you chose to have it here.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

File: instat/dlgUnstack.vb (hunk 4 above)
Comment (Gioche6) [reply instat/dlgUnstack.vb:129]: Done

File: instat/dlgUnstack.vb (hunk 6 above)
Comment (Gioche6) [reply instat/dlgUnstack.vb:366]: The default value is set to be 0, which don't appear as its the ucrNud default value. Hence I have made some changes and resulted adding the parameter manually for the 0 to appear in the` values_fill` parameter

**Comment by Gioche6** [comment]
@shadrackkibet , This is ready for review.

**Comment by Gioche6** [comment]
@rdstern . This is now ready for testing. Thank you.

**Comment by rdstern** [comment]
@Gioche6 the multiple option still seems flaky:
image
 The `columns to carry` control closes each time you return to the dialogue and I could not always get it to work when it had the data.  

I used the World Bank  data from the tidyR package.  You can unstack each year, so multiple years gives an example for multiple.  You need to carry the country.

**Comment by Gioche6** [comment]
@rdstern , I have made sure the checkbox works well now. Could you please test it. Thank you

**Review by rdstern — CHANGES_REQUESTED** [review]
@Gioche6 now I see what I did wrong last time.  But it is still not quite right - at least for beginners like me!
So with the same dataset I used multiple and then made the factor of the variable to unstack by.  Then I worked in order - please try too.  
I unstacked the data for 2000 to 2003.  Then I clicked on the Carry checkbox.  It now opens the multiple receiver, but it still leaves the focus on the columns to unstack!
So if you now click on Country, then it goes into the wrong receiver.  You don't see that, because the receiver is full.
Now you realise and it goes into the Carry receiver too.  But I don't think it works well when Country is in both!
The solution is simply when you click on the Carry checkbox, then move the focus automatically to that multiple receiver.

**Comment by Gioche6** [comment]
@rdstern , I have fixed the issue arising when changing the data frame and having the carry columns  receiver highlighted when the the carry columns checkbox is checked. Please test if it is working fine . Thank you

@shadrackkibet , this is ready for review.

**Review by rdstern — APPROVED** [review]
Great.  Seems to work.  Over to you @shadrackkibet

File: instat/dlgUnstack.vb [hunk 7]
```diff
@@ -41,118 +42,204 @@ Public Class dlgUnstack
...
-        'ucrMultipleColumnsReceiver.SetParameter(New RParameter("value.var", 1))
-        'ucrMultipleColumnsReceiver.SetParameterIsString()
-        'ucrMultipleColumnsReceiver.Selector = ucrSelectorForUnstack
+        'ucrMultipleColumnsReceiver
+        ucrMultipleColumnsReceiver.Selector = ucrSelectorForUnstack
+
+        ucrChkCarryColumns.SetText("Columns to Carry:")
+        ucrChkCarryColumns.SetParameter(New RParameter("checked", iNewPosition:=0))
+        ucrChkCarryColumns.SetValuesCheckedAndUnchecked("TRUE", "FALSE")
+        ucrChkCarryColumns.SetRDefault("FALSE")
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:75]: I understand this is a dummy function. I suggest you add this parameter in setDefaults then remove the RDefaults here.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

File: instat/dlgUnstack.vb (hunk 7 above)
Comment (Gioche6) [reply instat/dlgUnstack.vb]: This is Done

**Comment by Gioche6** [comment]
@shadrackkibet , I have resolved the issue above, Could you please see if its okay? Thank you.

**Comment by shadrackkibet** [comment]
Could you try to sort out the following minor issues;

1) OK should not be enabled in the following instances because when I click Ok I get bugs!

image
image

2) The checkbox appears to overlap with the receiver.

image

3) We shouldn't have instances where the current receiver is hidden. Otherwise, the user will be confused about where to add a variable.

image
4) I would prefer an input textbox rather than NUD for the following;
 
image

In the function help it is stated as follows;

>Optionally, a (scalar) value that specifies what each value should be filled in with when missing.
> This can be a named list if you want to apply different aggregations to different value columns.

You can ignore point (4) for now.

**Comment by Gioche6** [comment]
@shadrackkibet , I have resolved item 1 to 3. I have made sure the design is correctly implemented as well as Test Ok. Kindly review, Thank you.

**Comment by shadrackkibet** [comment]
@Gioche6 this is making good progress. Let us try to test it a bit more to identify if there are any bugs.

I tried to trick it a bit and I was able to get the following bug under Restore Hierarchy (Sorry I can't remember the steps) - By the way, I was using the French translation.
image

**Comment by Gioche6** [comment]
@shadrackkibet , I managed to detect the error and have now solved it. Additionally, I have worked on `TestOk` for the `rdo hierachy`. Kindly review and test. Thank you

**Comment by Gioche6** [comment]
@shadrackkibet , Could you review this. I've been testing using various datasets and works fine.

**Comment by shadrackkibet** [comment]
@Gioche6 thanks for the reminder. I will look into this now.

**Comment by shadrackkibet** [comment]
@Gioche6 I got the following still. 

1) Do we want to enable Ok when the multiple receiver is empty? I clicked Ok which produced the following code.

```
us_rent_income <- data_book$get_data_frame(data_name="us_rent_income")
us_rent_income_unstacked <- tidyr::pivot_wider(data=us_rent_income, names_from=NAME, values_from=moe)
data_book$import_data(data_tables=list(us_rent_income_unstacked=us_rent_income_unstacked))

rm(list=c("us_rent_income_unstacked", "us_rent_income"))
```
image

2) When I click single Ok is now disabled.

image

NB: A user can be easily confused and not know what to do when Ok is disabled in instances when it should be enabled.

**Comment by shadrackkibet** [comment]
@rdstern I don't remember why we choose to restrict this dialog to only unstack by factors. I think we should at least allow characters as well? this will also mean we change the label. What do you think?

**Comment by Gioche6** [comment]
@shadrackkibet , I have fixed the minor TESTOK bug. Kindly check if its okay. Thank you.

**Review by rdstern — CHANGES_REQUESTED** [review]
I have used the dialogue quite a lot in writing the help and it seems to be working well.  Just 2 small points:
a) When the data frame is changed the suggested name of the new file remains the same.  It should update.
b) One section not updated yet is the restore hierarchy options.  This still uses the reshape2::dcast function.  There is a data.table::dcast which is still current.  I found this worked just as well, so suggest that trivial change could be made.  Perhaps someone who knows more R, e.g. @shadrackkibet could confirm.

**Comment by shadrackkibet** [comment]
@Gioche6 could you sort out conflicts?

a) Straight forward to change.
b) I agree with the change to `data.table::dcast()`  for efficiency purposes. data.table functions are well known for that!

**Comment by Gioche6** [comment]
@Shadrack, I have sort out the merge conflicts and did some minor design changes. Kindly review. Thanks

**Comment by shadrackkibet** [comment]
> I have used the dialogue quite a lot in writing the help and it seems to be working well. Just 2 small points: a) When the data frame is changed the suggested name of the new file remains the same. It should update. b) One section not updated yet is the restore hierarchy options. This still uses the reshape2::dcast function. There is a data.table::dcast which is still current. I found this worked just as well, so suggest that trivial change could be made. Perhaps someone who knows more R, e.g. @shadrackkibet could confirm.

Did you also sort out this ones?

**Comment by Gioche6** [comment]
Item (a) is sorted. I am working on item (b)

**Comment by Gioche6** [comment]
@rdstern . I have changed the the function used in Restore Hierarchy to `data.table::dcast` as you suggested. I have also made sure that the Data frame names update. Kindly test. Thank you. 

@shadrackkibet , Kindly review. Thank you.

**Comment by shadrackkibet** [comment]
@rdstern this looks stable and complete now. Please test.

**Comment by shadrackkibet** [comment]
@rdstern this PR needs your assessment then we can possibly merge it.

**Comment by rdstern** [comment]
@shadrackkibet sorry, but the restore hierarchy now works slightly differently to the one before (in version 7.3.1)
image

Here it is in 7.3.1 from restore hierarchy with rwanda data from the Instat climatic library.  
image


Restore by Station name.  (Don't worry about the 8 lines rather than 4 - that's a different (interesting) question!)


It works in the pull request, but gives the results differently - including the counts for each station.  And doesn't include the unstacking variable.  That's different.  If I want those counts then I'll get them as a summary later!  

Could you adjust the command in the new version to give the results above
And it doesn't include

**Review by rdstern — CHANGES_REQUESTED** [review]
Change suggested to give same results as before.

**Comment by Gioche6** [comment]
@rdstern , I have made the adjustment needed. The restore hierarchy now produces the same result as before. I would love to know how you think about it after testing. Thank you
@shadrackkibet , This is ready for review. Thank you

**Comment by rdstern** [comment]
@Gioche6 that's better, because it now includes all variables.
image

But it still includes the counts which we don't need.  I hope that is also easy to fix.  I would like to get this merged in the coming version.

**Comment by Gioche6** [comment]
@rdstern , I have corrected that feature. The counts will no longer appear. Kindly test. Thank you

**Review by rdstern — APPROVED** [review]
@Gioche6 great.  I hope @lloyddewit and/or @shadrackkibet are ok with the code, so this can be merged.

**Comment by Gioche6** [comment]
@shadrackkibet , This is ready for review

**Comment by shadrackkibet** [comment]
- I am not sure what happened to the grid. See below

image

image

- I also get this when I try importing shapefiles.
image

image

This is an issue that might not be linked to this PR.

**Review by rdstern — DISMISSED** [review]
@shadrackkibet your comment made me check the dialogue a bit more - the other options.
I don't get your problem with the data sets I have used.  I agree it is probably a reogrid problem and not necessarily linked to this dialogue.  I have had your problem when (like your example) a variable contains a very long list.  It then seems to wordwrap - which I would sometimes like.  But in our case we don't yet have that option, so the cells then become very deep.  And as the word-wrapping is caused by a single variable - the polygon coordinates usually - it makes the grid look very odd.  Good to get some examples and make a separate issue I think.

File: instat/dlgUnstack.vb [hunk 8]
```diff
@@ -168,60 +258,125 @@ Public Class dlgUnstack
...
         TestOKEnabled()
     End Sub
+
+    Private Sub ucrChkValuesFill_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkValuesFill.ControlValueChanged, ucrNudValuesFill.ControlValueChanged
+        If ucrChkValuesFill.Checked Then
+            clsDcastFunction.AddParameter(New RParameter("values_fill", ucrNudValuesFill.GetText, iNewPosition:=7))
+        Else
+            clsDcastFunction.RemoveParameterByName("values_fill")
+        End If
+    End Sub
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:373]: I am not sure why you resorted to doing this manually anyway. In principle `ucrNudValuesFill`  should pass the parameter to `clsDcastFunction` function, then the linking you have in line 83 will take care of removing and adding? 
Does this make sense?

File: instat/dlgUnstack.vb [hunk 9]
```diff
@@ -168,60 +258,125 @@ Public Class dlgUnstack
...
             ucrReceiverCarryColumns.Location = New Point(ucrReceiverCarryColumns.Location.X, iReceiverMaxY)
+            ucrChkCarryColumns.Location = New Point(ucrChkCarryColumns.Location.X, iCarryClomunsY)
         End If
+
     End Sub
 
-    Private Sub ucrCoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrNewDFName.ControlContentsChanged, ucrReceiverFactorToUnstackby.ControlContentsChanged, ucrReceiverCarryColumns.ControlContentsChanged
+    Private Sub ucrCoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrNewDFName.ControlContentsChanged, ucrReceiverFactorToUnstackby.ControlContentsChanged, ucrReceiverCarryColumns.ControlContentsChanged, ucrNudValuesFill.ControlContentsChanged, ucrInputTextPrefix.ControlContentsChanged, ucrChkAddPrefix.ControlContentsChanged, ucrChkValuesFill.ControlContentsChanged, ucrMultipleColumnsReceiver.ControlContentsChanged, ucrReceiverColumnToUnstack.ControlContentsChanged, ucrChkCarryColumns.ControlContentsChanged
         TestOKEnabled()
     End Sub
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:365]: I always like making this the last `Sub` for consistency.

File: instat/dlgUnstack.vb [hunk 10]
```diff
@@ -168,60 +258,125 @@ Public Class dlgUnstack
...
+        If ucrChkValuesFill.Checked Then
+            clsDcastFunction.AddParameter(New RParameter("values_fill", ucrNudValuesFill.GetText, iNewPosition:=7))
+        Else
+            clsDcastFunction.RemoveParameterByName("values_fill")
+        End If
+    End Sub
+
+    Private Sub ucrReceiverColumnToUnstack_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverColumnToUnstack.ControlValueChanged
+        ValuesfillParameter()
+    End Sub
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:377]: Could you move this to line 188?

File: instat/dlgUnstack.vb [hunk 11]
```diff
@@ -168,60 +258,125 @@ Public Class dlgUnstack
...
+
+    Private Sub ucrReceiverFactorToUnstackby_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverFactorToUnstackby.ControlValueChanged
+        SetFormula()
+    End Sub
```
Comment (shadrackkibet) [inline instat/dlgUnstack.vb:381]: Perhaps even this one?

**Review by shadrackkibet — CHANGES_REQUESTED** [review]
@Gioche6 I had a few changes to request. Perhaps with a question.

**Comment by Gioche6** [comment]
@shadrackkibet , I have made the changes you requested.

**Review by rdstern — APPROVED** [review]
Still looks fine and a good improvement

**Review by shadrackkibet — APPROVED** [review]

**Pull request merged**